src/WebJobs.Script.WebHost/FileMonitoringService.cs

```diff
@@ -369,5 +382,41 @@ internal static async Task SetAppOfflineState(string rootPath, bool offline)
                 }, maxRetries: 3, retryInterval: TimeSpan.FromSeconds(1));
             }
         }
+
+        private class SuspendRestartRequest : IDisposable
+        {
+            private FileMonitoringService _fileMonitoringService;
+
+            public SuspendRestartRequest(FileMonitoringService fileMonitoringService)
+            {
+                _fileMonitoringService = fileMonitoringService;
+                if (Interlocked.Increment(ref _fileMonitoringService._suspensionRequestsCount) == 1)
+                {
+                    _fileMonitoringService._typedLogger.LogDebug("Restart is suspended");
+                }
+            }
+
+            public void Dispose()
+            {
+                if (Interlocked.Decrement(ref _fileMonitoringService._suspensionRequestsCount) == 0)
+                {
+                    _fileMonitoringService._typedLogger.LogDebug("Restart is resumed");
+                    if (Interlocked.Read(ref _fileMonitoringService._restartRequested) == 1)
+                    {
+                        _fileMonitoringService._typedLogger.LogDebug("Initiate force restart as it was triggerd during the suspension");
+                        // Blocking call as we can't use async call in Dispose method
+                        _fileMonitoringService._scriptHostManager.RestartHostAsync().GetAwaiter().GetResult();
+                    }
+                }
+                else
+                {
+                    // Error case
+                    if (_fileMonitoringService._suspensionRequestsCount < 0)
```

Comment:
Should we even allow that?

Reply:
I actually do not see how this error case can be reached, I will remove the code in next iteration.